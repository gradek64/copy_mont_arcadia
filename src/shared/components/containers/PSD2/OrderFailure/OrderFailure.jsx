import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { browserHistory } from 'react-router'

// Selectors
import { getLocationQuery } from '../../../../selectors/routingSelectors'
import { isUserAtLeastPartiallyAuthenticated } from '../../../../selectors/userAuthSelectors'

// Actions
import { getAccount } from '../../../../actions/common/accountActions'
import { setOrderError } from '../../../../actions/common/orderActions'
import { sendAnalyticsPaymentMethodPurchaseFailureEvent } from '../../../../analytics'

@connect(
  (state) => ({
    locationQuery: getLocationQuery(state),
    isAuthenticated: isUserAtLeastPartiallyAuthenticated(state),
  }),
  {
    setOrderError,
    sendAnalyticsPaymentMethodPurchaseFailureEvent,
  }
)
class OrderFailure extends Component {
  static needs = [getAccount]

  static propTypes = {
    locationQuery: PropTypes.object.isRequired,
    setOrderError: PropTypes.func.isRequired,
    sendAnalyticsPaymentMethodPurchaseFailureEvent: PropTypes.func.isRequired,
  }

  static contextTypes = {
    l: PropTypes.func,
  }

  componentDidMount() {
    const {
      sendAnalyticsPaymentMethodPurchaseFailureEvent,
      setOrderError,
      locationQuery: { error, orderId, paymentMethod },
      isAuthenticated,
    } = this.props

    if (error) {
      setOrderError(error)
    }

    sendAnalyticsPaymentMethodPurchaseFailureEvent({
      orderId,
      selectedPaymentMethod: paymentMethod,
    })

    if (!isAuthenticated) {
      browserHistory.replace('/guest/checkout/payment')
    } else {
      browserHistory.replace('/checkout')
    }
  }

  render() {
    const { l } = this.context

    return (
      <div className="OrderFailure">{l`Your order could not be completed`}</div>
    )
  }
}

export default OrderFailure
