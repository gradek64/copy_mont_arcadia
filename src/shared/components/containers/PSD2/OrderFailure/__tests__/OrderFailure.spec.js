import OrderFailure from '../OrderFailure'
import testComponentHelper from 'test/unit/helpers/test-component'

jest.mock('react-router', () => ({
  browserHistory: {
    replace: jest.fn(),
  },
}))
import { browserHistory } from 'react-router'

const setOrderError = jest.fn()
const sendAnalyticsPaymentMethodPurchaseFailureEvent = jest.fn()

describe('<OrderFailure />', () => {
  const renderComponent = testComponentHelper(OrderFailure.WrappedComponent)

  beforeEach(() => {
    jest.resetAllMocks()
  })

  describe('@renders', () => {
    it('displays a failure message', () => {
      const props = {
        locationQuery: {},
        setOrderError,
        sendAnalyticsPaymentMethodPurchaseFailureEvent,
      }

      expect(renderComponent(props).getTree()).toMatchSnapshot()
    })
  })

  describe('@componentDidMount', () => {
    it('redirects the client to /checkout', () => {
      const props = {
        locationQuery: {},
        setOrderError,
        sendAnalyticsPaymentMethodPurchaseFailureEvent,
        isAuthenticated: true,
      }

      const { instance } = renderComponent(props)

      expect(browserHistory.replace).not.toHaveBeenCalled()
      instance.componentDidMount()
      expect(browserHistory.replace).toHaveBeenCalledTimes(1)
      expect(browserHistory.replace).toHaveBeenCalledWith('/checkout')
    })

    it('redirects the client to /guest/checkout/payment for a guest user', () => {
      const props = {
        locationQuery: {},
        setOrderError,
        sendAnalyticsPaymentMethodPurchaseFailureEvent,
        isAuthenticated: false,
      }

      const { instance } = renderComponent(props)

      expect(browserHistory.replace).not.toHaveBeenCalled()
      instance.componentDidMount()
      expect(browserHistory.replace).toHaveBeenCalledTimes(1)
      expect(browserHistory.replace).toHaveBeenCalledWith(
        '/guest/checkout/payment'
      )
    })

    it('should call setOrderError() to capture an error passed as a query parameter', () => {
      const error = 'test-error'
      const props = {
        locationQuery: {
          error,
        },
        setOrderError,
        sendAnalyticsPaymentMethodPurchaseFailureEvent,
      }

      const { instance } = renderComponent(props)

      expect(setOrderError).not.toHaveBeenCalled()
      instance.componentDidMount()
      expect(setOrderError).toHaveBeenCalledTimes(1)
      expect(setOrderError).toHaveBeenCalledWith(error)
    })

    it('should call sendAnalyticsPaymentMethodPurchaseFailureEvent() to report to Google Analytics', () => {
      const orderId = 'test-order-id'
      const paymentMethod = 'test-payment-method'

      const props = {
        locationQuery: {
          orderId,
          paymentMethod,
        },
        setOrderError,
        sendAnalyticsPaymentMethodPurchaseFailureEvent,
      }

      const { instance } = renderComponent(props)

      expect(
        sendAnalyticsPaymentMethodPurchaseFailureEvent
      ).not.toHaveBeenCalled()
      instance.componentDidMount()
      expect(
        sendAnalyticsPaymentMethodPurchaseFailureEvent
      ).toHaveBeenCalledTimes(1)
      expect(
        sendAnalyticsPaymentMethodPurchaseFailureEvent
      ).toHaveBeenCalledWith({
        orderId,
        selectedPaymentMethod: paymentMethod,
      })
    })
  })
})
