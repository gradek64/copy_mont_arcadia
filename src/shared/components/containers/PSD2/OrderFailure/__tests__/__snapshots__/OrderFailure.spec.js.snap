// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderFailure /> @renders displays a failure message 1`] = `
<div
  className="OrderFailure"
>
  Your order could not be completed
</div>
`;
